### PR TITLE
[CARBONDATA-268]Improve carbonoptimizer Performance

### PR DESCRIPTION
**Pull request by QiangCai**
1. fix bug for carbon table union/join hive table

    No need to add dictionary decoder for hive table

2. just lookup relation from catalog once for each carbonrelation

3. use map for searching, instead of list

### CODE REVIEW

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala [hunk 1]
```diff
@@ -87,330 +130,363 @@ class ResolveCarbonFunctions(relations: Seq[CarbonDecoderRelation])
...
+
+  private def findRelation(attr: Attribute,
+       attrMap: java.util.Map[AttributeReferenceWrapper,
+       CarbonDecoderRelation]): Option[CarbonDecoderRelation] = {
+    val r = attrMap.get(AttributeReferenceWrapper(attr))
+    if(r == null) {
+      None
+    } else {
+      Some(r)
+    }
```
Comment (ravipesala) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala]: replace with Option(r) instead of `if else`

**Comment by ravipesala** [comment]
LGTM. Good work.
I can see lot of format changes, can you please avoid the format changes so that we can review the exact changes.

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala (hunk 1 above)
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala]: fixed

**Comment by QiangCai** [comment]
@ravipesala 
I have refactor code  and clean format changes.

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala [hunk 2]
```diff
@@ -87,20 +130,32 @@ class ResolveCarbonFunctions(relations: Seq[CarbonDecoderRelation])
...
     // collect alias information before hand.
     collectInformationOnAttributes(plan, aliasMap)
-    val transFormedPlan =
-      plan transformDown {
-        case cd: CarbonDictionaryTempDecoder if cd.isOuter =>
-          decoder = true
-          cd
+
+    def hasCarbonRelation(currentPlan: LogicalPlan): Boolean = {
+      val extraNodeInfo = mapOfNodes.get(currentPlan)
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala]: Change the name of mapofNodes to mapOfNonCarbonPlanNodes

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala [hunk 3]
```diff
@@ -68,9 +70,50 @@ object CarbonOptimizer {
...
   extends Rule[LogicalPlan] with PredicateHelper {
-
+  val LOGGER = LogServiceFactory.getLogService(this.getClass.getName)
   def apply(plan: LogicalPlan): LogicalPlan = {
-    transformCarbonPlan(plan, relations)
+    if (relations.nonEmpty && !isOptimized(plan)) {
+      LOGGER.info("Starting to optimize plan")
+      val startTime = System.currentTimeMillis
+      val result = transformCarbonPlan(plan, relations)
+      LOGGER.info("Time taken to optimize plan: " + ( System.currentTimeMillis - startTime))
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala]: change the statement to "Time taken for Carbon Optimizer to optimize"

Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala]: add this to STATISTICS logger type

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala [hunk 4]
```diff
@@ -397,20 +452,38 @@ class ResolveCarbonFunctions(relations: Seq[CarbonDecoderRelation])
         case others => others
       }
 
+    }
+
+    val tStartTime = System.currentTimeMillis
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonOptimizer.scala]: this variable not used

**Comment by QiangCai** [comment]
fixed comments

**Comment by gvramana** [comment]
LGTM

**Comment by gvramana** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/326/

**Pull request closed without merging**